File: go/triangle/triangle.go
package triangle

type Kind int

const (
	Equ Kind = iota // equilateral
	Iso             // isosceles
	Sca             // scalene
	NaT             // not a triangle
)

func KindFromSides(a, b, c float64) Kind {
	switch {
	case a < 0 || b < 0 || c < 0:
		return NaT
	case a >= b+c || b >= a+c || c >= a+b:
		return NaT
	case a == b && b == c:
		return Equ
	case a == b || a == c || b == c:
		return Iso
	default:
		return Sca
	}
}

File: cpp/triangle/triangle.cpp
#include <stdexcept>
#include "triangle.h"

namespace triangle {

triangle kind(double a, double b, double c) {
    if (a <= 0 || b <= 0 || c <= 0) {
        throw std::domain_error("invalid lenghts");
    }
    if (a + b < c || a + c < b || b + c < a) {
        throw std::domain_error("invalid lenght sum");
    }

    if (a == b && b == c) {
        return equilateral;
    } else if (a == b || a == c || b == c) {
        return isosceles;
    }
    return scalene;
}

}
